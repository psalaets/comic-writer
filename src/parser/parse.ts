import * as perf from '../perf';
import * as parts from '../comic-part-types';
import countWords from './count-words';
import * as classifiers from './line-classifiers';
import {
  Spread,
  Panel,
  Paragraph,
  Metadata,
  Dialogue,
  Caption,
  Sfx,
  LetteringContentChunk,
  BlankLine,
  PreSpread,
  SpreadChild,
  PanelChild,
  SpreadContent,
} from './types';

import {
  SPREAD_REGEX,
  PANEL_REGEX,
  CAPTION_REGEX,
  SFX_REGEX,
  DIALOGUE_REGEX,
  METADATA_REGEX,
  LETTERING_BOLD_REGEX
} from './regexes';

// flyweight pattern: reuse this to represent every blank line
const BLANK_LINE: BlankLine = {
  type: parts.BLANK
};

export const parseSpreadContent = perf.wrap('parseSpreadContent', parse);

function parse(content: SpreadContent): Spread<SpreadChild> {
  const spreadLine = content.spread;
  const childLines = content.children;

  const children = parseSpreadChildren(childLines);

  return parseSpread(spreadLine, children);
}

function parseSpreadChildren(lines: Array<string>): Array<SpreadChild> {
  const numbering = createLetteringNumberer();

  const spreadChildren: Array<SpreadChild> = [];
  let currentPanel: Panel<PanelChild> | null = null;

  for (let i = 0; i < lines.length; i++) {
    const parsed = parseSpreadChild(lines[i], numbering);

    if (parsed.type === parts.PANEL) {
      currentPanel = parsed;
      spreadChildren.push(parsed);
    } else {
      if (currentPanel) {
        currentPanel.children.push(parsed);
      } else {
        spreadChildren.push(parsed);
      }
    }
  }

  return applyPanelRollups(spreadChildren);
}

function* createLetteringNumberer() {
  let number = 1;
  while (true) {
    yield number++;
  }
}

function applyPanelRollups(children: Array<SpreadChild>): Array<SpreadChild> {
  children.forEach(spreadChild => {
    if (spreadChild.type === parts.PANEL) {
      const panel = spreadChild;

      panel.children.forEach(panelChild => {
        switch (panelChild.type) {
          case parts.CAPTION: {
            panel.captionCount += 1;
            panel.captionWordCount += panelChild.wordCount;
            break;
          }
          case parts.DIALOGUE: {
            panel.dialogueCount += 1;
            panel.dialogueWordCount += panelChild.wordCount;
            panel.speakers.push(panelChild.speaker);
            break;
          }
          case parts.SFX: {
            panel.sfxCount += 1;
            break;
          }
        }
      });
    }
  });

  return children;
}

function parseSpreadChild(line: string, numbering: Generator<number>): SpreadChild {
  // scripts are about 1/2 blank lines so this should be first
  if (classifiers.isBlank(line)) return BLANK_LINE;

  if (classifiers.isCaption(line)) return parseCaption(line, numbering);
  if (classifiers.isSfx(line)) return parseSfx(line, numbering);
  // dialogue has to be checked after sfx/caption, otherwise we get balloons
  // where the speaker is "caption" and "sfx"
  if (classifiers.isDialogue(line)) return parseDialogue(line, numbering);

  if (classifiers.isPanel(line)) return parsePanel(line);

  // any non-blank line can be a paragraph so it goes last
  return parseParagraph(line);
}

export function parsePreSpreadLines(lines: Array<string>): Array<PreSpread> {
  return lines.map(line => parsePreSpreadLine(line));
}

function parsePreSpreadLine(line: string): PreSpread {
  // scripts are about 1/2 blank lines so this should be first
  if (classifiers.isBlank(line)) return BLANK_LINE;

  if (classifiers.isMetadata(line)) return parseMetadata(line);

  // any non-blank line can be a paragraph so it goes last
  return parseParagraph(line);
}

export function parseSpread(line: string, children: Array<SpreadChild>): Spread<SpreadChild> {
  const matchResult = SPREAD_REGEX.exec(line) as Array<string>;

  const startPage = Number(matchResult[1]);
  const endPage = matchResult[3] != null ? Number(matchResult[3]) : startPage;
  const pageCount = countPages(startPage, endPage);

  return applySpreadRollups({
    type: parts.SPREAD,
    pageCount,
    children,

    // these start with default values, they're populated when applying rollups
    panelCount: 0,
    speakers: [],
    dialogueCount: 0,
    captionCount: 0,
    sfxCount: 0,
    dialogueWordCount: 0,
    captionWordCount: 0
  });
}

function applySpreadRollups(spread: Spread<SpreadChild>): Spread<SpreadChild> {
  return spread.children.reduce((spread, child) => {
    switch (child.type) {
      case parts.PANEL:
        spread.panelCount        += 1;
        spread.captionCount      += child.captionCount;
        spread.captionWordCount  += child.captionWordCount;
        spread.dialogueCount     += child.dialogueCount;
        spread.dialogueWordCount += child.dialogueWordCount;
        spread.sfxCount          += child.sfxCount;
        spread.speakers.push(...child.speakers);
        break;
      case parts.DIALOGUE:
        spread.dialogueCount     += 1;
        spread.dialogueWordCount += child.wordCount;
        break;
      case parts.CAPTION:
        spread.captionCount      += 1;
        spread.captionWordCount  += child.wordCount;
        break;
    }

    return spread;
  }, spread);
}

function countPages(startPage: number, endPage?: number): number {
  if (endPage == null) {
    return 1;
  } else if (startPage < endPage) {
    return (endPage - startPage) + 1;
  } else if (startPage > endPage) {
    return 2;
  } else { // startPage === endPage
    return 1;
  }
}

export function parsePanel(line: string): Panel<PanelChild> {
  const [, number] = PANEL_REGEX.exec(line) as Array<string>;

  return {
    type: parts.PANEL,
    number: Number(number),

    // these start with default values, they'll be populated later
    children: [],
    speakers: [],
    dialogueCount: 0,
    captionCount: 0,
    sfxCount: 0,
    dialogueWordCount: 0,
    captionWordCount: 0,
  };
}

export function parseMetadata(line: string): Metadata {
  const [, name, value] = METADATA_REGEX.exec(line) as Array<string>;

  return {
    type: parts.METADATA,
    name,
    value,
  };
}

export function parseParagraph(line: string): Paragraph {
  return {
    type: parts.PARAGRAPH,
    content: line
  };
}

export function parseDialogue(line: string, numbering: Generator<number>): Dialogue {
  const [, speaker, modifier, content] = DIALOGUE_REGEX.exec(line) as Array<string>;

  const parseTree = parseLetteringContent(content);

  return {
    type: parts.DIALOGUE,
    number: numbering.next().value,
    speaker,
    modifier: modifier ? modifier.slice(1, -1) : null,
    content: parseTree,
    wordCount: countWords(parseTree),
  };
}

export function parseCaption(line: string, numbering: Generator<number>): Caption {
  const [, modifier, content] = CAPTION_REGEX.exec(line) as Array<string>;
  const parseTree = parseLetteringContent(content);

  return {
    type: parts.CAPTION,
    number: numbering.next().value,
    modifier: modifier ? modifier.slice(1, -1) : null,
    content: parseTree,
    wordCount: countWords(parseTree)
  };
}

export function parseSfx(line: string, numbering: Generator<number>): Sfx {
  const [, modifier, content] = SFX_REGEX.exec(line) as Array<string>;

  return {
    type: parts.SFX,
    number: numbering.next().value,
    modifier: modifier ? modifier.slice(1, -1) : null,
    content,
  };
}

function parseLetteringContent(content: string): Array<LetteringContentChunk> {
  const chunks: Array<LetteringContentChunk> = [];

  let index = 0;
  let result = null;

  // eslint-disable-next-line no-cond-assign
  while (result = LETTERING_BOLD_REGEX.exec(content.slice(index))) {
    const before = content.slice(index, index + result.index)

    if (before) {
      chunks.push({
        type: parts.TEXT,
        content: before
      });
    }

    chunks.push({
      type: parts.LETTERING_BOLD,
      content: result[1]
    })

    index += result.index + result[0].length;
  }

  const after = content.slice(index);
  if (after) {
    chunks.push({
      type: parts.TEXT,
      content: after
    });
  }

  return chunks;
}
