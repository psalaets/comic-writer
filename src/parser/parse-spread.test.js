import { parseSpread } from './parse';

describe('parseSpread()', () => {
  it('single page', () => {
    const result = parseSpread(lines(`Page 1`));

    expect(result).toMatchSnapshot();
  });

  it('2 page range', () => {
    const result = parseSpread(lines(`Pages 1-2`));

    expect(result).toMatchSnapshot();
  });

  it('3 page range', () => {
    const result = parseSpread(lines(`Pages 1-3`));

    expect(result).toMatchSnapshot();
  });

  it('inverted page range', () => {
    const result = parseSpread(lines(`Pages 2-1`));

    expect(result).toMatchSnapshot();
  });

  it('self to self page range', () => {
    const result = parseSpread(lines(`Pages 2-2`));

    expect(result).toMatchSnapshot();
  });

  it('page with panels', () => {
    const result = parseSpread(lines(`Page 1

Panel 1

Panel 2`));

    expect(result).toMatchSnapshot();
  });

  it('page with panels and lettering', () => {
    const result = parseSpread(lines(`Page 1

Panel 1

\tbob: ok

\tbob: try it again

Panel 2

This is panel description.

\tsfx: boom

\tcaption: yep`));

    expect(result).toMatchSnapshot();
  });

  it('dialogue with modifier', () => {
    const result = parseSpread(lines(`Page 1
Panel 1

\tbob (yell): go!`));

    expect(result).toMatchSnapshot();
  });

  it('dialogue with bold', () => {
    const result = parseSpread(lines(`Page 1
Panel 1

\tbob: eat **this** not **that**!

\tbob: this is *not*`));

    expect(result).toMatchSnapshot();
  });

  it('dialogue starts with bold', () => {
    const result = parseSpread(lines(`Page 1
Panel 1

\tbob: **only bold**

\tbob: **starts bold** and then
`));

    expect(result).toMatchSnapshot();
  });

  it('caption with bold', () => {
    const result = parseSpread(lines(`Page 1
Panel 1

\tcaption: **I** didn't mean **that**!

\tcaption: this is *not*`));

    expect(result).toMatchSnapshot();
  });

  it('kitchen sink', () => {
    const result = parseSpread(lines(`Pages 1-2

Panel 1

Some characters stand around. And do stuff.

\tbob: ok

\tbob: try it **again**

Panel 2

This is panel description.

\tsfx: boom

\tcaption: yep

Panel 3

And they walk off into the sunset.
`));

    expect(result).toMatchSnapshot();
  });

  it('first line is not a page', () => {
    expect(() => {
      parseSpread(lines('Panel 1'));
    }).toThrowError();
  });
});

function lines(source) {
  return source.split('\n');
}
