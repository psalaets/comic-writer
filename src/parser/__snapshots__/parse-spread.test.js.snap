// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseSpread() 2 page range 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 2,
  "panelCount": 0,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;

exports[`parseSpread() 3 page range 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 3,
  "panelCount": 0,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;

exports[`parseSpread() caption with bold 1`] = `
Object {
  "captionCount": 2,
  "captionWordCount": 7,
  "content": Array [
    Object {
      "captionCount": 2,
      "captionWordCount": 7,
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": "I",
              "type": "lettering-bold",
            },
            Object {
              "content": " didn't mean ",
              "type": "text",
            },
            Object {
              "content": "that",
              "type": "lettering-bold",
            },
            Object {
              "content": "!",
              "type": "text",
            },
          ],
          "lineOffset": 3,
          "modifier": null,
          "type": "caption",
          "wordCount": 4,
        },
        Object {
          "content": Array [
            Object {
              "content": "this is *not*",
              "type": "text",
            },
          ],
          "lineOffset": 5,
          "modifier": null,
          "type": "caption",
          "wordCount": 3,
        },
      ],
      "dialogueCount": 0,
      "dialogueWordCount": 0,
      "lineOffset": 1,
      "sfxCount": 0,
      "speakers": Array [],
      "type": "panel",
    },
  ],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 1,
  "panelCount": 1,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;

exports[`parseSpread() dialogue starts with bold 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": "only bold",
              "type": "lettering-bold",
            },
          ],
          "lineOffset": 3,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 2,
        },
        Object {
          "content": Array [
            Object {
              "content": "starts bold",
              "type": "lettering-bold",
            },
            Object {
              "content": " and then",
              "type": "text",
            },
          ],
          "lineOffset": 5,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 4,
        },
      ],
      "dialogueCount": 2,
      "dialogueWordCount": 6,
      "lineOffset": 1,
      "sfxCount": 0,
      "speakers": Array [
        "bob",
        "bob",
      ],
      "type": "panel",
    },
  ],
  "dialogueCount": 2,
  "dialogueWordCount": 6,
  "pageCount": 1,
  "panelCount": 1,
  "sfxCount": 0,
  "speakers": Array [
    "bob",
    "bob",
  ],
  "type": "spread",
}
`;

exports[`parseSpread() dialogue with bold 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": "eat ",
              "type": "text",
            },
            Object {
              "content": "this",
              "type": "lettering-bold",
            },
            Object {
              "content": " not ",
              "type": "text",
            },
            Object {
              "content": "that",
              "type": "lettering-bold",
            },
            Object {
              "content": "!",
              "type": "text",
            },
          ],
          "lineOffset": 3,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 4,
        },
        Object {
          "content": Array [
            Object {
              "content": "this is *not*",
              "type": "text",
            },
          ],
          "lineOffset": 5,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 3,
        },
      ],
      "dialogueCount": 2,
      "dialogueWordCount": 7,
      "lineOffset": 1,
      "sfxCount": 0,
      "speakers": Array [
        "bob",
        "bob",
      ],
      "type": "panel",
    },
  ],
  "dialogueCount": 2,
  "dialogueWordCount": 7,
  "pageCount": 1,
  "panelCount": 1,
  "sfxCount": 0,
  "speakers": Array [
    "bob",
    "bob",
  ],
  "type": "spread",
}
`;

exports[`parseSpread() dialogue with modifier 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": "go!",
              "type": "text",
            },
          ],
          "lineOffset": 3,
          "modifier": "yell",
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 1,
        },
      ],
      "dialogueCount": 1,
      "dialogueWordCount": 1,
      "lineOffset": 1,
      "sfxCount": 0,
      "speakers": Array [
        "bob",
      ],
      "type": "panel",
    },
  ],
  "dialogueCount": 1,
  "dialogueWordCount": 1,
  "pageCount": 1,
  "panelCount": 1,
  "sfxCount": 0,
  "speakers": Array [
    "bob",
  ],
  "type": "spread",
}
`;

exports[`parseSpread() inverted page range 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 2,
  "panelCount": 0,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;

exports[`parseSpread() kitchen sink 1`] = `
Object {
  "captionCount": 1,
  "captionWordCount": 1,
  "content": Array [
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [
        Object {
          "content": "Some characters stand around. And do stuff.",
          "lineOffset": 4,
          "type": "paragraph",
        },
        Object {
          "content": Array [
            Object {
              "content": "ok",
              "type": "text",
            },
          ],
          "lineOffset": 6,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 1,
        },
        Object {
          "content": Array [
            Object {
              "content": "try it ",
              "type": "text",
            },
            Object {
              "content": "again",
              "type": "lettering-bold",
            },
          ],
          "lineOffset": 8,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 3,
        },
      ],
      "dialogueCount": 2,
      "dialogueWordCount": 4,
      "lineOffset": 2,
      "sfxCount": 0,
      "speakers": Array [
        "bob",
        "bob",
      ],
      "type": "panel",
    },
    Object {
      "captionCount": 1,
      "captionWordCount": 1,
      "content": Array [
        Object {
          "content": "This is panel description.",
          "lineOffset": 12,
          "type": "paragraph",
        },
        Object {
          "content": "boom",
          "lineOffset": 14,
          "modifier": null,
          "type": "sfx",
        },
        Object {
          "content": Array [
            Object {
              "content": "yep",
              "type": "text",
            },
          ],
          "lineOffset": 16,
          "modifier": null,
          "type": "caption",
          "wordCount": 1,
        },
      ],
      "dialogueCount": 0,
      "dialogueWordCount": 0,
      "lineOffset": 10,
      "sfxCount": 1,
      "speakers": Array [],
      "type": "panel",
    },
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [
        Object {
          "content": "And they walk off into the sunset.",
          "lineOffset": 20,
          "type": "paragraph",
        },
      ],
      "dialogueCount": 0,
      "dialogueWordCount": 0,
      "lineOffset": 18,
      "sfxCount": 0,
      "speakers": Array [],
      "type": "panel",
    },
  ],
  "dialogueCount": 2,
  "dialogueWordCount": 4,
  "pageCount": 2,
  "panelCount": 3,
  "sfxCount": 1,
  "speakers": Array [
    "bob",
    "bob",
  ],
  "type": "spread",
}
`;

exports[`parseSpread() page with panels 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [],
      "dialogueCount": 0,
      "dialogueWordCount": 0,
      "lineOffset": 2,
      "sfxCount": 0,
      "speakers": Array [],
      "type": "panel",
    },
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [],
      "dialogueCount": 0,
      "dialogueWordCount": 0,
      "lineOffset": 4,
      "sfxCount": 0,
      "speakers": Array [],
      "type": "panel",
    },
  ],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 1,
  "panelCount": 2,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;

exports[`parseSpread() page with panels and lettering 1`] = `
Object {
  "captionCount": 1,
  "captionWordCount": 1,
  "content": Array [
    Object {
      "captionCount": 0,
      "captionWordCount": 0,
      "content": Array [
        Object {
          "content": Array [
            Object {
              "content": "ok",
              "type": "text",
            },
          ],
          "lineOffset": 4,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 1,
        },
        Object {
          "content": Array [
            Object {
              "content": "try it again",
              "type": "text",
            },
          ],
          "lineOffset": 6,
          "modifier": null,
          "speaker": "bob",
          "type": "dialogue",
          "wordCount": 3,
        },
      ],
      "dialogueCount": 2,
      "dialogueWordCount": 4,
      "lineOffset": 2,
      "sfxCount": 0,
      "speakers": Array [
        "bob",
        "bob",
      ],
      "type": "panel",
    },
    Object {
      "captionCount": 1,
      "captionWordCount": 1,
      "content": Array [
        Object {
          "content": "This is panel description.",
          "lineOffset": 10,
          "type": "paragraph",
        },
        Object {
          "content": "boom",
          "lineOffset": 12,
          "modifier": null,
          "type": "sfx",
        },
        Object {
          "content": Array [
            Object {
              "content": "yep",
              "type": "text",
            },
          ],
          "lineOffset": 14,
          "modifier": null,
          "type": "caption",
          "wordCount": 1,
        },
      ],
      "dialogueCount": 0,
      "dialogueWordCount": 0,
      "lineOffset": 8,
      "sfxCount": 1,
      "speakers": Array [],
      "type": "panel",
    },
  ],
  "dialogueCount": 2,
  "dialogueWordCount": 4,
  "pageCount": 1,
  "panelCount": 2,
  "sfxCount": 1,
  "speakers": Array [
    "bob",
    "bob",
  ],
  "type": "spread",
}
`;

exports[`parseSpread() self to self page range 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 1,
  "panelCount": 0,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;

exports[`parseSpread() single page 1`] = `
Object {
  "captionCount": 0,
  "captionWordCount": 0,
  "content": Array [],
  "dialogueCount": 0,
  "dialogueWordCount": 0,
  "pageCount": 1,
  "panelCount": 0,
  "sfxCount": 0,
  "speakers": Array [],
  "type": "spread",
}
`;
